pkg/controllers/user/monitoring/config_refresh_handler.go

```diff
@@ -0,0 +1,112 @@
+package monitoring
+
+import (
+	"encoding/json"
+	"reflect"
+	"sort"
+
+	"github.com/rancher/rancher/pkg/controllers/user/nslabels"
+	apiv1 "github.com/rancher/types/apis/core/v1"
+	rmonitoringv1 "github.com/rancher/types/apis/monitoring.coreos.com/v1"
+
+	monitoringv1 "github.com/coreos/prometheus-operator/pkg/client/monitoring/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	projectNSAnnotation = "project.cattle.io/namespaces"
+)
+
+type ConfigRefreshHandler struct {
+	prometheusClient rmonitoringv1.PrometheusInterface
+	prometheusLister rmonitoringv1.PrometheusLister
+	nsClient         apiv1.NamespaceInterface
+}
+
+func (h *ConfigRefreshHandler) syncNamespace(key string, obj *corev1.Namespace) (runtime.Object, error) {
+	if obj == nil || obj.DeletionTimestamp != nil {
+		return obj, nil
+	}
+	promList, err := h.prometheusLister.List("", labels.NewSelector())
+	if err != nil {
+		return obj, err
+	}
+
+	for _, prometheus := range promList {
```

Comment:
Im really concerned about enqueuing all prom operators every time a namspace gets sync. In a cluster with a lot of projects and a lot of namespaces, this can get expensive, especially since you are using clients instead of listers `syncPrometheus`, those always go all the way to etcd. You are also doing json unmarshalling, which is expensive and resource intensive.

Can you make this logic more selective such you only get prometheus CRDs that you care about? That would be any CRDs it is currently a part of and the one it needs to be a part of.
I think you can achieve this using an indexer for the prometheus CRD. In the indexer, parse your namespace annotation to return each namespace as a key.

Reply:
I changed `nsClient` to `nsLister`. It would be less expensive. I can't not make this more selective because in the namespace moving from `p1` to `p2` case, we can't know the original project the namespace belongs to. 

So I need to regenerate the project-namespace relation every time to ensure prometheus crd for `p1` get updated when ns is changed. And the operator will refresh the `p1` prometheus configuration and the old namespace scrape configuration will be moved.

Reply:
You can know all the prometheus objects that have the namespace as an object.
Index function would look something like this:
```
func promsByMemberNamespace(obj object) {
    // 1. parse the json annotation containing all the namesapces
    namespaces = doParsing(obj)

    return []string{namespaces}  // just pseudo code showing that im returning all the namespaces for this prom object
   
}
...
func syncNamespace(obj namesapce) {
   // get all prometheus objects that are currently scraping this NS
   prometheusObjects = indexer.GetByKey("prom-by-member-namespace", obj.Name)
   
   for (p := range promethuesObjects) {
      enqueue(p)
   }
}
````